drivers/pinctrl/pinctrl-bflb.c

```diff
@@ -0,0 +1,802 @@
+		if (arg) {
+			bflb_gpio_set_reg
+			(
+				pctl,
+				offset,
+				REG_GPIOx_PD | REG_GPIOx_PU,
+				FIELD_PREP(REG_GPIOx_PD, 0) | FIELD_PREP(REG_GPIOx_PU, 0)
+			);
+		}
+		break;
+
+	case PIN_CONFIG_BIAS_PULL_DOWN:
+		if (arg) {
+			bflb_gpio_set_reg
+			(
+				pctl,
+				offset,
+				REG_GPIOx_PD | REG_GPIOx_PU,
+				FIELD_PREP(REG_GPIOx_PD, 1) | FIELD_PREP(REG_GPIOx_PU, 0)
+			);
+		}
+		break;
+
+	case PIN_CONFIG_BIAS_PULL_UP:
+		if (arg) {
+			bflb_gpio_set_reg
+			(
+				pctl,
+				offset,
+				REG_GPIOx_PD | REG_GPIOx_PU,
+				FIELD_PREP(REG_GPIOx_PD, 0) | FIELD_PREP(REG_GPIOx_PU, 1)
+			);
+		}
+		break;
+
+	case PIN_CONFIG_BIAS_HIGH_IMPEDANCE:
+		bflb_gpio_set_reg
+		(
+			pctl,
+			offset,
+			REG_GPIOx_PD | REG_GPIOx_PU | REG_GPIOx_IE | REG_GPIOx_OE,
+			FIELD_PREP(REG_GPIOx_PD, 0) | FIELD_PREP(REG_GPIOx_PU, 0) |
+				FIELD_PREP(REG_GPIOx_IE, 0) | FIELD_PREP(REG_GPIOx_OE, 0)
+		);
+		break;
+
+	case PIN_CONFIG_INPUT_ENABLE:
+		bflb_gpio_set_reg
+		(
+			pctl,
+			offset,
+			REG_GPIOx_IE,
+			FIELD_PREP(REG_GPIOx_IE, !!arg)
+		);
+		break;
+
+	case PIN_CONFIG_INPUT_SCHMITT_ENABLE:
+		bflb_gpio_set_reg
+		(
+			pctl,
+			offset,
+			REG_GPIOx_SMT,
+			FIELD_PREP(REG_GPIOx_SMT, !!arg)
+		);
+		break;
+
+	case PIN_CONFIG_OUTPUT:
+		bflb_gpio_set_reg
+		(
+			pctl,
+			offset,
+			REG_GPIOx_OE | REG_GPIOx_O,
+			FIELD_PREP(REG_GPIOx_OE, 1) | FIELD_PREP(REG_GPIOx_O, !!arg)
+		);
+		break;
+
+	case PIN_CONFIG_OUTPUT_ENABLE:
+		bflb_gpio_set_reg
+		(
+			pctl,
+			offset,
+			REG_GPIOx_OE,
+			FIELD_PREP(REG_GPIOx_OE, 1)
+		);
+		break;
+
+	default: return -ENOTSUPP;
+	}
+
+	dev_dbg(pctl->dev, "Pin %u config set to %lu (param %u, arg %u)", offset,
+		config, param, arg);
+
+	return 0;
+}
+
+/* IRQ chip functions */
+
+//AH: Clear the interrupt for the specified GPIO
+static void bflb_gpio_irq_ack(struct irq_data *data)
+{
+	struct bflb_gpio_pinctrl *pctl =
+			gpiochip_get_data(irq_data_get_irq_chip_data(data));
+
+	bflb_gpio_set_reg
+	(
+		pctl,
+		data->hwirq,
+		REG_GPIOx_INT_CLR,
+	    FIELD_PREP(REG_GPIOx_INT_CLR, 1)
+	);
+
+	bflb_gpio_set_reg
+	(
+		pctl,
+		data->hwirq,
+		REG_GPIOx_INT_CLR,
+	    FIELD_PREP(REG_GPIOx_INT_CLR, 0)
+	);
+
+	dev_dbg(pctl->dev, "Pin %lu IRQ ACK", data->hwirq);
+}
+
+//AH: Find the correct value for the type of interrupts we want to receive
+//for a GPIO
+static unsigned int bflb_gpio_irq_type(unsigned int type)
+{
+	switch (type & IRQ_TYPE_SENSE_MASK) {
+
+	case IRQ_TYPE_EDGE_RISING: return BFLB_IRQ_MODE_SYNC_EDGE_RISING;
+
+	case IRQ_TYPE_EDGE_FALLING: return BFLB_IRQ_MODE_SYNC_EDGE_FALLING;
+
+	case IRQ_TYPE_EDGE_BOTH: return BFLB_IRQ_MODE_SYNC_EDGE_BOTH;
+
+	case IRQ_TYPE_LEVEL_HIGH: return BFLB_IRQ_MODE_SYNC_LEVEL_HIGH;
+
+	case IRQ_TYPE_LEVEL_LOW: return BFLB_IRQ_MODE_SYNC_LEVEL_LOW;
+
+	//No "off" available on BL808, set to default IRQ_TYPE_EDGE_FALLING and
+	//then we'll need to mask
+	default: return BFLB_IRQ_MODE_SYNC_EDGE_FALLING;
+	}
+
+	return BFLB_IRQ_MODE_SYNC_EDGE_FALLING;
```

Comment:
nit: use a local variable to capture the return type and you can avoid having multiple points of exit from this function, and this unreachable code here